### PR TITLE
Levels feature 3

### PR DESCRIPTION
**Pull request by AidanFromProgramming**
Added moderation related features to levels along with some additional catches to prevent excessive xp

### CODE REVIEW

**Review by tweirtx — DISMISSED** [review]

File: dozer/cogs/levels.py [hunk 1]
```diff
@@ -44,9 +44,12 @@ def total_xp_for_level(level):
         # https://github.com/Mee6/Mee6-documentation/blob/9d98a8fe8ab494fd85ec27750592fc9f8ef82472/docs/levels_xp.md
         # > The formula to calculate how many xp you need for the next level is 5 * (lvl ^ 2) + 50 * lvl + 100 with
         # > your current level as lvl
+        if level >= 1000000:  # If the level gets too big, dozer will hang trying to calculate the level. A better way needs to be found to calculate
+            DOZER_LOGGER.warning("Member XP exceeded maximum calculation limit")  # level's but this is the best for now
```
Comment (bkeeneykid) [inline dozer/cogs/levels.py:48]: This should be a critical error, not a warning, as this is breaking functionality of the bot.

File: dozer/cogs/levels.py [hunk 2]
```diff
@@ -59,7 +62,7 @@ def level_for_total_xp(xp):
         # > The formula to calculate how many xp you need for the next level is 5 * (lvl ^ 2) + 50 * lvl + 100 with
         # > your current level as lvl
         lvl = 0
-        while xp >= 0:
+        while xp >= 0 and lvl <= 1000000:  # The same limitation is applied here
```
Comment (bkeeneykid) [inline dozer/cogs/levels.py:65]: If `lvl` is indeed over 1000000 here, this function would end up returning -1, which seems like unintended behavior.

File: dozer/cogs/levels.py [hunk 3]
```diff
@@ -317,6 +334,82 @@ async def checkrolelevels(self, ctx):
...
+    `{prefix}adjustlevels swapxp <@Snowplow or "Snowplow"> <@Dozer or "Dozer">`:\n Swaps Snowplow's xp with Dozer's xp
+    `{prefix}adjustlevels transferxp <@Snowplow or "Snowplow"> <@Dozer or "Dozer">`:\n Adds Snowplow's xp to dozer's xp
+    """
+
+    @adjustlevels.command()
+    @guild_only()
+    @has_permissions(manage_messages=True)
+    async def setlevel(self, ctx, member: discord.Member, level: int):
+        """Changes a members level to requested level"""
+        if level >= 100000:
```
Comment (bkeeneykid) [inline dozer/cogs/levels.py:355]: This max level which is hard coded in several places in this new code should really be a cog-wide constant.

File: dozer/cogs/levels.py [hunk 4]
```diff
@@ -317,6 +334,82 @@ async def checkrolelevels(self, ctx):
...
+        e.add_field(name='Success!', value=f"I set {member}'s level to {level}")
+        e.set_footer(text='Triggered by ' + ctx.author.display_name)
+        await ctx.send(embed=e)
+
+    @adjustlevels.command(aliases=["addxp"])
+    @guild_only()
+    @has_permissions(manage_messages=True)
+    async def changexp(self, ctx, member: discord.Member, xp_amount: int):
+        """Changes a members xp by a certain amount"""
+        if abs(xp_amount) >= 2147483647:
```
Comment (bkeeneykid) [inline dozer/cogs/levels.py:372]: This should also be a cog-wide constant. In addition, there should be a significant bumper under this XP amount such that the XP amount doesn't get increased above this limit in other functions.

File: dozer/cogs/levels.py [hunk 5]
```diff
@@ -317,6 +334,82 @@ async def checkrolelevels(self, ctx):
...
+        e = discord.Embed(color=blurple)
+        e.add_field(name='Success!', value=f"I set {member}'s level to {level}")
+        e.set_footer(text='Triggered by ' + ctx.author.display_name)
+        await ctx.send(embed=e)
+
+    @adjustlevels.command(aliases=["addxp"])
+    @guild_only()
+    @has_permissions(manage_messages=True)
+    async def changexp(self, ctx, member: discord.Member, xp_amount: int):
+        """Changes a members xp by a certain amount"""
```
Comment (bkeeneykid) [inline dozer/cogs/levels.py:371]: This description is misleading. Changing XP implies that this command sets the member's XP to that level, while this function actually adds it to the current total XP.

File: dozer/cogs/levels.py [hunk 6]
```diff
@@ -317,6 +334,82 @@ async def checkrolelevels(self, ctx):
...
+
+    @adjustlevels.command()
+    @guild_only()
+    @has_permissions(manage_messages=True)
+    async def transferxp(self, ctx, take_member: discord.Member, give_member: discord.Member):
+        """Adds xp from one member to another member"""
+        take = await self.load_member(ctx.guild.id, take_member.id)
+        give = await self.load_member(ctx.guild.id, give_member.id)
+        give.total_xp += take.total_xp
+        give.total_messages += take.total_messages
```
Comment (bkeeneykid) [inline dozer/cogs/levels.py:406]: Should this function not set the `take` member's XP to zero?

**Review by bkeeneykid — CHANGES_REQUESTED** [review]

**Review by bkeeneykid — APPROVED** [review]

**Review by tweirtx — APPROVED** [review]

**Pull request merged**